ENH run_in_pyodide type annotations, defaults, posonly, and kwonly args

hoodmane opened the pull request:
It still doesn't support *args or **kwargs arguments though.

- [x] Add tests
- [ ] Update changelog

ryanking13 commented:
Thanks for keep improving `run_in_pyodide`! For the part about modifying the AST, it's hard to understand if it's correct just by looking at the code... so I'll review it again when the tests are added.

hoodmane commented:
Okay I added tests. Of course it didn't work very well before I added the tests.

I reorganized things a bit to reduce the reliance on class attributes when it can be avoided and make things a bit easier to read. Unfortunately this made the diff a bit big.

ryanking13 commented on tests/test_decorator.py:
```diff
@@ -296,3 +298,57 @@ def helper(selenium, meth):
+    def fn_with_posonly(selenium, a, /, b):
+        return None
+
+    fn_with_posonly(selenium, 5, 7)
```
Could you also check the values allocated a and b are correct?

ryanking13 reviewed (approved):
Thanks! I should admit that I don't fully understand the AST hacks. Anyway, the test passes so I am okay to merge this.
